c2rust-analyze/src/main.rs

```diff
@@ -35,29 +41,113 @@ mod pointer_id;
 mod type_desc;
 mod util;
 
+/// A wrapper around `T` that dynamically tracks whether it's initialized or not.  `RefCell`
+/// dynamically tracks borrowing and panics if the rules are violated at run time; `MaybeUnset`
+/// dynamically tracks initialization and similarly panics if the value is accessed while unset.
+#[derive(Clone, Copy, Debug)]
+struct MaybeUnset<T>(Option<T>);
+
+impl<T> Default for MaybeUnset<T> {
+    fn default() -> MaybeUnset<T> {
+        MaybeUnset(None)
+    }
+}
+
+impl<T> MaybeUnset<T> {
+    pub fn set(&mut self, x: T) {
+        if self.0.is_some() {
+            panic!("value is already set");
+        }
+        self.0 = Some(x);
+    }
+
+    pub fn clear(&mut self) {
+        if self.0.is_none() {
+            panic!("value is already cleared");
+        }
+        self.0 = None;
```

Comment:
```suggestion
        if self.0.take().is_none() {
            panic!("value is already cleared");
        }
```

Reply:
IMO this makes the logic harder to follow

Reply:
I think we also get a better, more standard, more recognizable and related to `Option` error message here with `expect`.

Reply:
I hadn't considered the panic message.  IMO, the fact that the message would mention `Option` is a (very minor) downside of the `expect` version.  The user of `MaybeUnset<T>` should not be thinking of it as an `Option<T>`.  They should think of it as a `T`, with the minor caveat that Rust's normal static checks for variable initialization checks don't apply.

Reply:
Here I see `take` as in it's taking away the current value, the same as clearing it.

Reply:
This simply is not an idiomatic use of `take`.  The primary purpose of `take` is to get the old value.  This is reflected in the docs ("Takes the value out of the option") and in common usage.  I did a quick grep of rust-lang/rust, and every instance of `take()`, `take().unwrap()`, or `take().expect()` that I found (outside of test code) was actually using the extracted value for something afterwards.